### PR TITLE
Add separate nsg for the router

### PR DESCRIPTION
**Pull request by kargakis**
Fixes https://github.com/kargakis/acs-engine/issues/62

### CODE REVIEW

**Comment by kargakis** [comment]
@pweil- @jim-minter got a separate nsg working for the router. I had to create a separate subnet because the master nsg is linked to the main subnet and its rules apply to every resource inside that subnet.

Still WIP, needs cleanup

**Comment by jim-minter** [comment]
@kargakis I don't know what the tradeoffs/recommendations are around different subnets per agent type vs attaching different nsgs directly to different VMs sharing a single subnet.
@jackfrancis please could you weigh in on this / pull in others as appropriate?  The use case here is opening up ports on the infra VMs, which we do not want to open on the master or compute VMs.

**Comment by pweil-** [comment]
> had to create a separate subnet because the master nsg is linked to the main subnet and its rules apply to every resource inside that subnet.

The requirement to support existing vnets makes me assume that creating a new subnet for this is not the way we should be going.  

> attaching different nsgs directly to different VMs sharing a single subnet

We can attach to the subnet or NIC through resource manager.  NIC sounds like the way we're trying to go.

**Comment by kargakis** [comment]
> NIC sounds like the way we're trying to go.

Yeah, I was thinking of exploring this route today.

**Comment by kargakis** [comment]
Pulling this out of WIP; moved the master nsg to apply for the master nic.

File: parts/k8s/kubernetesmasterresources.t [hunk 1]
```diff
@@ -1,4 +1,42 @@
 {{if IsOpenShift}}
+    {
+      "apiVersion": "[variables('apiVersionDefault')]",
+      "location": "[variables('location')]",
+      "name": "router-nsg",
```
Comment (jim-minter) [inline parts/k8s/kubernetesmasterresources.t]: I think this name should be templatised to end up looking like the others, also "infra", not "router" - so it turns out something like ocp-infra-30982460-nsg

**Comment by kargakis** [comment]
I can't see any nsg attached into the master nic; looking

File: parts/k8s/kubernetesagentresourcesvmas.t [hunk 2]
```diff
@@ -5,16 +5,22 @@
         "name": "loop"
       },
       "dependsOn": [
+{{if .IsOpenShiftInfra }}
+      "[resourceId('Microsoft.Network/networkSecurityGroups', 'router-nsg')]",
+{{end}}
 {{if .IsCustomVNET}}
-      "[variables('nsgID')]"
-{{else}}
```
Comment (jim-minter) [inline parts/k8s/kubernetesagentresourcesvmas.t]: this change doesn't look right to me

File: parts/k8s/kubernetesagentresourcesvmas.t [hunk 3]
```diff
@@ -5,16 +5,22 @@
...
       ],
       "location": "[variables('location')]",
       "name": "[concat(variables('{{.Name}}VMNamePrefix'), 'nic-', copyIndex(variables('{{.Name}}Offset')))]",
       "properties": {
-{{if .IsCustomVNET}}
+{{if .IsOpenShiftInfra }}
+        "networkSecurityGroup": {
+          "id": "[resourceId('Microsoft.Network/networkSecurityGroups', 'router-nsg')]"
+        },
+{{else if .IsCustomVNET}}
```
Comment (jim-minter) [inline parts/k8s/kubernetesagentresourcesvmas.t]: Don't .IsOpenShiftInfra and .IsCustomVNET need to be orthogonal?  Can this be done?  Is it necessary to define a second thing like nsgID?

File: parts/k8s/kubernetesagentresourcesvmas.t (hunk 2 above)
Comment (kargakis) [reply parts/k8s/kubernetesagentresourcesvmas.t]: Right, I thought I was fixing a template bug here but it seems I am wrong

File: parts/k8s/kubernetesmasterresources.t (hunk 1 above)
Comment (kargakis) [reply parts/k8s/kubernetesmasterresources.t]: agreed; we need to templatize all of the router resources. Maybe a follow-up?

File: parts/k8s/kubernetesagentresourcesvmas.t (hunk 3 above)
Comment (kargakis) [reply parts/k8s/kubernetesagentresourcesvmas.t]: In an Openshift installation, in any case (custom or not) we want an infra VM to use the router nsg, right?

Comment (pweil-) [reply parts/k8s/kubernetesagentresourcesvmas.t]: yes, we have to set up the same nsg structure in a custom vnet.

File: parts/k8s/kubernetesagentresourcesvmas.t (hunk 2 above)
Comment (jim-minter) [reply parts/k8s/kubernetesagentresourcesvmas.t]: ignore these comments, am going to delete them and have a second go at reviewing this ;)

**Comment by kargakis** [comment]
This is ready for another round of reviews

**Comment by jim-minter** [comment]
This looks good to me; I'd like to see
- (hopefully) rubberstamp from @jackfrancis
- follow-up issue opened to deal with naming new resources in line with existing ones, and adding things like `variables('infransgID')`, etc.

My impression is that it would make sense to set the `dependsOn`s such that the nics depend on the relevant nsg, and the nsgs depend on the vnet if !IsCustomVNET - but only if upstream agrees and it need not be in this PR.

**Comment by jim-minter** [comment]
@jackfrancis this PR wants to do two things - 1) move nsg attachment from the subnet to the nic; 2) allow infra node VMs to use a different nsg to master/agent VMs.  Quite potentially master/agent VMs might need separate nsgs in the future too.

**Comment by pweil-** [comment]
Looking good to me too, but man these golang templates are a pain to review.  

> Quite potentially master/agent VMs might need separate nsgs in the future too.

Agree, they should each have their own

File: parts/k8s/kubernetesmasterresources.t [hunk 4]
```diff
@@ -1,4 +1,42 @@
...
+              "direction": "Inbound",
+              "priority": 111,
+              "protocol": "Tcp",
+              "sourceAddressPrefix": "*",
+              "sourcePortRange": "*"
+            }
+          }
+        ]
+      },
+      "type": "Microsoft.Network/networkSecurityGroups"
```
Comment (jackfrancis) [inline parts/k8s/kubernetesmasterresources.t]: nit: move this up to be below the name so it's easier for ye olde humans to read

**Comment by pweil-** [comment]
> would you like that to happen against this fork, or should we open up a PR

I think we're very close to opening an upstream PR with our changes, let's run it upstream.  I think the question will be about organization of that commit so reviewers don't look at the commits and close the tab immediately. :smile:

**Comment by jim-minter** [comment]
Thanks @jackfrancis.  I've opened https://github.com/Azure/acs-engine/pull/2604, which I believe is a faithful representation of this PR, for backwards compatibility tests - do you think you could run these today?

File: parts/k8s/kubernetesmasterresources.t (hunk 4 above)
Comment (kargakis) [reply parts/k8s/kubernetesmasterresources.t]: done

**Comment by kargakis** [comment]
/hold

I prefer to merge this one after we merge https://github.com/Azure/acs-engine/pull/2611.

**Comment by jim-minter** [comment]
@kargakis I forgot to mention to you today, @jackfrancis would like these changes to be openshift-only because there's not currently the bandwidth to support applying the change to all k8s deployments.  @pweil- I believe the consensus remains that we're better off doing it this way than the multiple subnet way.

**Comment by pweil-** [comment]
> @pweil- I believe the consensus remains that we're better off doing it this way than the multiple subnet way.

:+1:

**Comment by kargakis** [comment]
Yeap, will open a PR today or tomorrow.

**Comment by kargakis** [comment]
> Yeap, will open a PR today or tomorrow.

Opened https://github.com/Azure/acs-engine/pull/2695

**Pull request closed without merging**